src/charm.py

```diff
@@ -201,8 +202,11 @@ def _install_or_upgrade(self, _event=None):
         self.unit.status = MaintenanceStatus("Deploying vSphere Cloud Provider")
         self.unit.set_workload_version("")
         for controller in self.collector.manifests.values():
-            controller.apply_manifests()
-        self.stored.deployed = True
+            try:
+                controller.apply_manifests()
+                self.stored.deployed = True
+            except ApiError:
+                self.unit.status = WaitingStatus("Waiting for kube-apiserver")
```

Comment:
What do you think about only setting `deployed=True` after the whole loop has finished.

if we have an APIError -- let's defer the event?

```suggestion
            try:
                controller.apply_manifests()
            except ApiError:
                self.unit.status = WaitingStatus("Waiting for kube-apiserver")
                _event.defer()
                return
        self.stored.deployed = True
```

Reply:
Yes, I think it's better to defer the event and setting the flag after the whole loop finishes. I will commit your suggestion right away.